stellargraph/mapper/graph_generator.py

```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from ..core.graph import StellarGraph
+from ..core.utils import is_real_iterable
+from .sequences import GraphSequence
+from ..core.experimental import experimental
+
+
+@experimental(reason="Missing unit tests.", issues=[1042])
+class GraphGenerator:
```

Comment:
On this one I would have to disagree. This generator is not specific and certainly it is not full batch. It generates batches of graphs which as you point out are represented by their adjacency and node feature matrices but it can be no other way since the input to neural networks has to be in tensor form. I specifically called this `GraphGenerator` because I hope we will be able to re-use in the future with other graph classification algorithms. For example, we should be able to re-use it as is for #923.

Reply:
My reasoning with calling it "specific" was that this generator yields the graphs in a specific format, not that it relates to a specific algorithm or can't be reused (neither of my suggested names hint at that, just like the `FullBatchNodeGenerator` type doesn't talk about a specific algorithm, but does hint at *how* it yields the graph). In particular, it pads all matrices to all have the same shape (given this, another name could be `PaddedGraphGenerator`).

My reasoning with the "full batch" name is that it yields the full information about a graph, and so is, in some sense, a generalisation of `FullBatchNodeGenerator` to have a batch size > 1. This could potentially be another (maybe long term) approach here, unify this with `FullBatch*Generator` by allowing passing multiple graphs into those types.

Other options I could see for types that could be called a "graph generator":

- a generator that only yields a single graph at a time (although I guess this would just be an optimised form of `.flow(..., batch_size=1)`)
- something using [ragged tensors](https://www.tensorflow.org/guide/ragged_tensor) to not have to do any padding

However, I guess this is a perfectly reasonable default behaviour one might expect from something called a "graph generator".

Reply:
Your definition of "full batch" is not standard. The data here is not a single graph but a collection of graphs. It would be full batch if we required that all graphs are processed at once for each epoch. This is not the case. For now, this generator is correctly named. We can revisit in the future if the need arises. I am not familiar with ragged tensors and trying to re-write the generator now to use them would not be the best use of my time considering that there is no guarantee that it will work as it is now.

Reply:
> Your definition of "full batch" is not standard. The data here is not a single graph but a collection of graphs. It would be full batch if we required that all graphs are processed at once for each epoch

You're right, that full-batch is a misleading name here (and hence it was only one of several that I provided), but I don't think any of your reply addresses the fact that what we call `FullBatchNodeGenerator` and what we call `GraphGenerator` are doing essentially the same thing, just the latter allows more than one graph as input. Their output is even passed into essentially the same type of model, for essentially the same task (at least, `FullBatchNodeGenerator` can be used for computing node embeddings).

Finding generalisations like this is a core part of both mathematics and programming...

> For now, this generator is correctly named. We can revisit in the future if the need arises

I'm not saying it is incorrectly named. I'm saying that it _may_ be _sub-optimally_ named, because it takes up a really basic name in "API space" and thus makes future generalisations difficult.

Like with the name of `GraphClassification`, revisiting in future is hard: if users already have a pile of code using `GraphGenerator`, it is extremely poor of us to rename that without some sort of migration assistance. Having that migration assistance likely means we cannot use `GraphGenerator` for something else, like a superclass for all graph generators.

> I am not familiar with ragged tensors and trying to re-write the generator now to use them would not be the best use of my time considering that there is no guarantee that it will work as it is now.

I don't think I was even hinting at doing any rewrites or changing the code.

I'm trying to think of options that we _might_ want, to explore whether we're going to have other forms of "graph generators". If we're definitely not going to have any other forms, then calling this singular one (that's the only one we'll ever have) `GraphGenerator` is fine. If there may be other forms, then we may want to think about the name. My list was a list of possibilities for other forms.

I'm trying to have a discussion about future possibilities here, to make sure we're not boxing ourselves into a suboptimal corner. If you're ok with the risk that we'll be stuck with `GraphGenerator` referring to this specific style of generating graphs, I guess we'll just make do with it as is.

Reply:
I'm happy to stop trying to think about this for now and just land it as is. We can have a longer term brainstorm, until the point we de-`@experimental`-ise it.

Reply:
I suggest that instead of discussing what-if scenarios and library-wide API design in any single ticket, we should, as you suggest, have a proper design discussion offline. There are many small differences between all the different generators, models, and layers that force us to replicate code. You are correct that `FullBatchGenerator` and `GraphGenerator` are similar and in fact, the latter is more general and could be used for training node representation models, if it wasn't for our "need" to output predictions only for a subset of the nodes, hence the need for `out_indices` and the call to `gather` in the end. So the way I see it, it is `FullBatchGenerator` that will eventually be deprecated for not being very flexible, not `GraphGenerator`.

Anyway, I think we agree to just table this discussion for now.